src/test/java/org/elasticsearch/test/engine/MockEngineSupport.java

```diff
@@ -0,0 +1,180 @@
+            this.wrapper = wrapper;
+            this.indexSettings = indexSettings;
+            flushOnClose = indexSettings.getAsDouble(FLUSH_ON_CLOSE_RATIO, 0.5d);
+        }
+    }
+
+    public static final ConcurrentMap<AssertingSearcher, RuntimeException> INFLIGHT_ENGINE_SEARCHERS = new ConcurrentHashMap<>();
+
+    private final MockContext mockContext;
+
+    public MockEngineSupport(EngineConfig config) {
+        Settings indexSettings = config.getIndexSettings();
+        final long seed = indexSettings.getAsLong(ElasticsearchIntegrationTest.SETTING_INDEX_SEED, 0l);
+        Random random = new Random(seed);
+        final double ratio = indexSettings.getAsDouble(WRAP_READER_RATIO, 0.0d); // DISABLED by default - AssertingDR is crazy slow
+        Class<? extends AssertingDirectoryReader> wrapper = indexSettings.getAsClass(READER_WRAPPER_TYPE, AssertingDirectoryReader.class);
+        boolean wrapReader = random.nextDouble() < ratio;
+        if (logger.isTraceEnabled()) {
+            logger.trace("Using [{}] for shard [{}] seed: [{}] wrapReader: [{}]", this.getClass().getName(), config.getShardId(), seed, wrapReader);
+        }
+        mockContext = new MockContext(random, wrapReader, wrapper, indexSettings);
+    }
+
+    enum CloseAction {
+        FLUSH_AND_CLOSE,
+        CLOSE;
+    }
+
+
+    /**
+     * Returns the CloseAction to execute on the actual engine. Note this method changes the state on
+     * the first call and treats subsequent calls as if the engine passed is already closed.
+     */
+    public CloseAction flushOrClose(Engine engine, CloseAction originalAction) throws IOException {
+        try {
+            if (closing.compareAndSet(false, true)) { // only do the random thing if we are the first call to this since super.flushOnClose() calls #close() again and then we might end up with a stackoverflow.
+                if (mockContext.flushOnClose > mockContext.random.nextDouble()) {
+                    return CloseAction.FLUSH_AND_CLOSE;
+                } else {
+                    return CloseAction.CLOSE;
+                }
+            } else {
+                return originalAction;
+            }
+        } finally {
+            if (logger.isTraceEnabled()) {
+                // log debug if we have pending searchers
+                for (Map.Entry<AssertingSearcher, RuntimeException> entry : INFLIGHT_ENGINE_SEARCHERS.entrySet()) {
+                    logger.trace("Unreleased Searchers instance for shard [{}]",
+                            entry.getValue(), entry.getKey().shardId());
+                }
+            }
+        }
+    }
+
+    public AssertingIndexSearcher newSearcher(Engine engine, String source, IndexSearcher searcher, SearcherManager manager) throws EngineException {
+        IndexReader reader = searcher.getIndexReader();
+        IndexReader wrappedReader = reader;
+        assert reader != null;
+        if (reader instanceof DirectoryReader && mockContext.wrapReader) {
```

Comment:
ok, i see it. Its just a little non-obvious due to the way searchers are bubbled up. maybe we can add an assert in the future.